evap/evaluation/auth.py

```diff
@@ -45,160 +48,101 @@ def authenticate(self, request, email=None, password=None):  # pylint: disable=a
         return None
 
 
-def user_passes_test(test_func):
-    """
-    Decorator for views that checks whether a user passes a given test
-    (raising 403 if not). The test should be a callable that takes the
-    user object and returns True if the user passes.
-    """
-
-    def decorator(view_func):
-        @wraps(view_func)
-        def _wrapped_view(request, *args, **kwargs):
+def class_or_function_check_decorator(test_func: Callable[[UserProfile], bool]):
+    def function_decorator(func):
+        @wraps(func)
+        def wrapped(request, *args, **kwargs):
             if not test_func(request.user):
-                raise PermissionDenied()
-            return view_func(request, *args, **kwargs)
+                raise PermissionDenied
+            return func(request, *args, **kwargs)
 
-        return _wrapped_view
+        return wrapped
 
-    return decorator
+    def decorator(class_or_function):
+        if inspect.isclass(class_or_function):
+            return method_decorator(function_decorator, name="dispatch")(class_or_function)
```

Comment:
Maybe add a comment pointing to https://docs.djangoproject.com/en/4.2/topics/class-based-views/intro/#decorating-the-class, and possibly explain what we're doing here, something like:
> Using django's `method_decorator` and hardcoding `name="dispatch"` so it can be applied to view classes. See https://docs.djangoproject.com/en/4.2/topics/class-based-views/intro/#decorating-the-class